### PR TITLE
Enchancement: to_raw_bytes(Variant) ->PackedByteArray: in GlobalScope

### LINKED ISSUE #7478: Convert an array into byte stream for SSBO and push_constant uniform fill out.
### Describe the project you are working on

Experimenting with compute shaders

### Describe the problem or limitation you are having in your project

Preparing PackedByteArray for SSBO or push constants fill-out is multistep, inconvenient, error prone and slow.
Maintaining consistent variable types and structures in GDScript and compute shaders  is a separate task anyway as there is no builtin functionality in Godot. This is OK for demos or experiments, but this is no go for large projects which you have to maintain.

Consider following parameters that you have to pass to SSBO of a compute shader (or to push_constant):
```GDScript
                 var params:Dictionary = { # Vector3 must be 16-bytes aligned.
		"camera_postion":Vector3(0,0,0),
		"light_energy":$DirectionalLight3D.light_energy, # last 4 bytes of alignment for previous vec3
		"light_direction":$DirectionalLight3D.global_transform.basis.z,
		"seed":int(seed_), # last 4 bytes of alignment for previous vec3
		"sun_disk_size":float(sun_disk_size),
		"phase_g":float(G),
		"clouds_density":float(clouds_density),
		"clouds_coverage":float(clouds_coverage),
		"clouds_low":float(clouds_low),
		"clouds_high":float(clouds_high),
		"weather_scale":float(weather_scale),
		"time":float(time),
		"time_scale":float(time_scale),
		"time_offset":float(time_offset)
	}
```

You can't convert it directly with var_to_bytes, because you will get 156 bytes instead of 72 which are required to fill the SSBO in compute shader, due to encoding. 

You need to convert all of these params into floats then push them into PackedFloat32Array, then convert it to PackedByteArray and then update your SSBO or push to constants uniform. 

Like this:

```GDScript
        var f32a:PackedFloat32Array = PackedFloat32Array()
        for k in params:
		match(typeof(params[k])):
			TYPE_VECTOR3:
				arr.push_back(params[k].x)
				arr.push_back(params[k].y)
				arr.push_back(params[k].z)
			TYPE_FLOAT:
				arr.push_back(params[k])
			TYPE_BOOL: # what value will be stored into float in SSBO if the value is 'true' ?  Ok, it will not be 0 so it will probably be true
				arr.push_back(float(params[k]))
			TYPE_INT:   # I wonder what happens here if in SSBO this is a uint value ... 
				arr.push_back(float(params[k]))
# for the push_constant you will probably need to add some more bytes at the end for alignment.
```

**So this makes a semantic conversion not the bit conversion, and may cause a lot of trouble and debugging.**

Maintaining the code for consistency between params in GDScript is a separate task anyways but this makes it even  more complicated and error prone.

And you do the conversion within GDScript, which is slow especially when you want to use low latency push_constant which suppose to be the fastest. This makes no sense at all. 

The better solution would be to introduce the "SSBO" class that acts as a Dictionary in GDScript, but supports limited amount of types. This can be done in GDScript though and if you inherit from Resource class then it will be serializable out of the box too. You can add a function that generates glsl.inc for your push_constant or SSBO, which partially solves the maintaining problem. 

So no need in additional functionality? 

Nope. 

**There is still a problem of semantic vs bit conversion.**

And performance problem still there but it can be discarded for now.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Introduce the to_raw_bytes() GlobalScope method which is able to convert bit-wise some limited amount of Variants (Array, int, float, bool, Vector2, Vector3,Vector4,Vector2i,Vector3i,Vector4i) . 

This PR solves the bit-wise conversion problem and simplifies the conversion to a byte stream from an Array. 
to_raw_bytes() does not convert the Dictionaries though (yet?). It does not implement anything that can be done in GDScript, but may come handy when you need to create "class_name SSBO extends Resource" which will do the params maintenance for you.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is the PR https://github.com/godotengine/godot/pull/80500

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Bit-wise conversion will be very complicated  in GDScript in this case. I'm not as advanced in GDScript to say that this is impossible, though I found no easy way here.

### Is there a reason why this should be core and not an add-on in the asset library?

Bit-wise conversion functionality should be the "first class support" not the external library or extension.

To improve performance one can consider making an SSBO/push_constant class as an extension/library in C++ instead of GDScript, though, which does not discards the requirement for bit-wise conversion functionality, especially for push_constant which must be low latency.

### PR DESCRIPTION
**Pull request by pkpro**
### What  
```GDScript
 to_raw_bytes(Variant) -> PackedByteArray
 ```
A new function in a GlobalScope to convert a Variant into PackedByteArray which does not include an encoding information (no type information is encoded, no possibility to de-serialize from result)

### Why
There is no simple way to prepare a byte stream for push constants or for SSBO updates of the compute shaders in GDScript.  It is inconvenient, slow and error prone.

Here is a simple function that convert limited range of Variants (Array, int, float, bool, Vector2, Vector3,Vector4,Vector2i,Vector3i,Vector4i) into byte stream and returns it as PackedByteArray.

Nope, this function does not control the alignment if vec2,vec3 types, this is still responsibility of the user.

Here how it looks like:

```GDScript
    # Vectors are aligned to 16 bytes, so we need to pad vectors with some other 4 byte value like float, int or bool
    var array = [
        Vector3(camera_pos.x,camera_pos.y,camera_pos.z),
        float(seed),                                                    # padded for 16 byte alignment of the camera_position above
        Vector3(light_direction.x,light_direction.y,light_direction.z),
        float(scale),                                                   # padded for 16 byte alignment of the light_direction above
        [ 0, 1, true, 33.17, Vector3(0.1,0.2,0.3), false ]              # some other array values, Vector3 is padded for 16 byte alignment
                                                                            # by bool value, and before Vector3 there are 4 four-byte values
    ]
    
    var bytes:PackedByteArray = to_raw_bytes(array)                     # convert to raw bytes
    _rd.buffer_update(_buffer_rid, 0, bytes.size(), bytes)                # update existing SSBO
```

### Bad

One should avoid using the result of this function with bytes_to_vars() function, because of the obvious possibility to not only get a garbage in return or impossibility to convert, but a danger of engine crash.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/7478*

### CODE REVIEW

**Comment by AThousandShips** [comment]
This warrants a [proposal](https://github.com/godotengine/godot-proposals) to discuss the specifics and the support for such a feature, make sure to fill in the form properly

File: core/io/marshalls.h [hunk 1]
```diff
@@ -215,6 +217,8 @@ class EncodedObjectAsID : public RefCounted {
 Error decode_variant(Variant &r_variant, const uint8_t *p_buffer, int p_len, int *r_len = nullptr, bool p_allow_objects = false, int p_depth = 0);
 Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bool p_full_objects = false, int p_depth = 0);
 
+Error to_raw_bytes(const Variant &p_variant, PackedByteArray &out_sink);
```
Comment (AThousandShips) [inline core/io/marshalls.h:220]:
```suggestion
Error to_raw_bytes(const Variant &p_variant, PackedByteArray &r_out_sink);
```
We use this prefix for return values

File: core/io/marshalls.cpp [hunk 2]
```diff
@@ -1814,6 +1814,82 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
 	return OK;
 }
 
+Error to_raw_bytes(const Variant &p_variant, PackedByteArray &out) {
```
Comment (AThousandShips) [inline core/io/marshalls.cpp:1817]:
```suggestion
Error to_raw_bytes(const Variant &p_variant, PackedByteArray &r_out_sink) {
```
Keep consistent

File: core/io/marshalls.cpp [hunk 3]
```diff
@@ -1814,6 +1814,82 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
...
+		case Variant::VECTOR4I: {
+			Vector4i v = p_variant;
+			for (size_t elements = 0; elements < 4; elements++) {
+				to_raw_bytes(v[elements], out);
+			}
+		} break;
+		case Variant::ARRAY: {
+			Array v = p_variant;
+			for (int i = 0; i < v.size(); i++) {
+				to_raw_bytes(v[i], out);
```
Comment (AThousandShips) [inline core/io/marshalls.cpp:1883]:
```suggestion
				Error err = to_raw_bytes(v[i], out);
				if (err) {
					return err;
				}
```
Need to report the error here, as the array can contain values that raise errors

File: core/io/marshalls.cpp [hunk 4]
```diff
@@ -1814,6 +1814,82 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
...
+			size_t offset = out.size();
+			out.resize(out.size() + sizeof(v));
+			memcpy(out.ptrw() + offset, &v, sizeof(v));
+#endif
+		} break;
+		case Variant::VECTOR2: {
+			Vector2 v = p_variant;
+			for (size_t elements = 0; elements < 2; elements++) {
+				to_raw_bytes(v[elements], out);
+			}
```
Comment (AThousandShips) [inline core/io/marshalls.cpp:1848]: This needs to be changed as per the comment about `float`

File: core/io/marshalls.cpp [hunk 5]
```diff
@@ -1814,6 +1814,82 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
...
+			double v = p_variant;
+			size_t offset = out.size();
+			out.resize(out.size() + sizeof(v));
+			memcpy(out.ptrw() + offset, &v, sizeof(v));
+#else
+			float v = p_variant;
+			size_t offset = out.size();
+			out.resize(out.size() + sizeof(v));
+			memcpy(out.ptrw() + offset, &v, sizeof(v));
+#endif
```
Comment (AThousandShips) [inline core/io/marshalls.cpp:1842]: This is not correct, `float` in `Variant` is always 64 bits, only the values of `Vector2/3/4` are 32/64 bits, this needs to be changed

File: core/io/marshalls.cpp [hunk 6]
```diff
@@ -1814,6 +1814,82 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
...
+			Vector4 v = p_variant;
+			for (size_t elements = 0; elements < 4; elements++) {
+				to_raw_bytes(v[elements], out);
+			}
+		} break;
+		case Variant::VECTOR2I: {
+			Vector2i v = p_variant;
+			for (size_t elements = 0; elements < 2; elements++) {
+				to_raw_bytes(v[elements], out);
+			}
```
Comment (AThousandShips) [inline core/io/marshalls.cpp:1866]: This is not correct either, the values of `Vector2/3/4i` are all 32 bits, whereas the `int` is always 64 bits, this needs to be handled specifically

**Review by AThousandShips — DISMISSED** [review]
Thank you for contributing 

Several issues, this needs to be rewritten based on this, and use more appropriate code IMO

You should probably use the `encode_` methods, like `encode_real`, taking into account details

**Comment by pkpro** [comment]
> This warrants a [proposal](https://github.com/godotengine/godot-proposals) to discuss the specifics and the support for such a feature, make sure to fill in the form properly
> 
> I fail to see the need for this, it can be easily worked around with the existing encode methods, and doing so gives you full freedom to control the alignment

It is as I wrote: It is not easy, inconvenient, slow and error prone.

I'll fill out the proposal now.

**Comment by AThousandShips** [comment]
You didn't mention the encode methods in `PackedByteArray` which is what I mean here, which allows a lot of flexibility, is easy, and not error prone IMO

**Comment by pkpro** [comment]
> Thank you for contributing
> 
> Several issues, this needs to be rewritten based on this, and use more appropriate code IMO
> 
> You should probably use the `encode_` methods, like `encode_real`, taking into account details

This is exactly what to_raw_bytes() have to avoid. See the proposal here in details: https://github.com/godotengine/godot-proposals/issues/7478

The problem with bit-wise conversion has no other solution than some trickery and workarounds in GDScript though, if any.

This is non serializable byte stream for the specific use with compute shaders params udpates.

File: core/io/marshalls.cpp (hunk 6 above)
Comment (pkpro) [reply core/io/marshalls.cpp:1866]: int and uint in GLSL are 32 bit integers, though you are correct about GDScript int's, they are 64 bits and this have to be adressed.

File: core/io/marshalls.cpp (hunk 5 above)
Comment (pkpro) [reply core/io/marshalls.cpp:1842]: float is 32 bits and only 64 bits when REAL_T_IS_DOUBLE is defined.. Variant, does not matter as there is no mapping from Variant to GLSL types. Vectors in GLSL are 16 bytes aligned and Vector2 for example will take 2xfloat - 8 bytes and 8 pad bytes. Padding is not the part of the job for this function.

Comment (AThousandShips) [reply core/io/marshalls.cpp:1842]: Float in GDScript is 64 bits

**Comment by AThousandShips** [comment]
Making this shader specific is not right, it should be general purpose if it is in `@GlobalScope`

File: core/io/marshalls.cpp [hunk 7]
```diff
@@ -1814,6 +1814,82 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
 	return OK;
 }
 
+Error to_raw_bytes(const Variant &p_variant, PackedByteArray &out) {
+	switch (p_variant.get_type()) {
+		case Variant::INT: {
+			int32_t v = p_variant;
```
[the remark this answers is not in the document]
Comment (pkpro) [reply core/io/marshalls.cpp:1820]: In GLSL uint and int are 32 bit, here goes the semantic conversion first. If value can't fit into 32 bits, so be it. However this must be documented that there are no 64 bit integers support in this method.

File: core/io/marshalls.cpp (hunk 3 above)
Comment (pkpro) [reply core/io/marshalls.cpp]: Added a commit to address these issues in raw_to_bytes(). There are a lot of places where  errors cases were ignored.

File: core/io/marshalls.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (pkpro) [reply core/io/marshalls.cpp:1820]: Added a documentation note with describing lack of 64 bit integers support for this method.

File: core/io/marshalls.cpp (hunk 5 above)
Comment (pkpro) [reply core/io/marshalls.cpp]: Removed use of the "double" completely, added semantic conversion to 32 bit float at a cost of precision.

File: core/io/marshalls.cpp (hunk 6 above)
Comment (pkpro) [reply core/io/marshalls.cpp:1875]: This is fixed with the commit [3266fe5f9da822baba14b2998cf9823405ebdfc1](https://github.com/godotengine/godot/commit/3266fe5f9da822baba14b2998cf9823405ebdfc1), by converting doubles and uint64_t of the GDScript into 32 bit floats and integers.
Documentation updated: [3e7b1bf9417c03c547bb12f802e14d24c2617287](https://github.com/godotengine/godot/commit/3e7b1bf9417c03c547bb12f802e14d24c2617287)

**Comment by pkpro** [comment]
> You didn't mention the encode methods in `PackedByteArray` which is what I mean here, which allows a lot of flexibility, is easy, and not error prone IMO
> 

I might be missing something here, care to elaborate?

> Your suggested approach doesn't allow any flexibility on precision or details like that, which is a drawback, it also doesn't allow you to pass `PackedInt*Array`, `PackedColorArray`, or `PackedFloat*Array`, or even `Color`
> 

You are correct in this regard. I didn't worked on this part yet as on the Dictionary conversion as well. I might extend it later though. However I'm not sure this proposal and the functionality will benefit someone beside me yet. So I acted early and we can decide if we go one with it or drop it completely. 


> Also I don't know if the example you provided is realistic, you will never send arbitrary data to a compute shader, you have to match the data format, so you can always use individual methods to do it, using the methods on the `PackedByteArray`, you should also know the size beforehand, so resizing the data to fit isn't really an issue either, IMO

There are a lot of scenarios to send arbitrary data to compute shaders, starting with any kind of procedural environment and ending with SFX, like rain drops or even SDF rendering.

> you should also know the size beforehand, so resizing the data to fit isn't really an issue either, IMO

Yes that is the game developer task and you can't address this with the code in other way but some simplification for the cost of precision for example.

**Comment by pkpro** [comment]
> Making this shader specific is not right, it should be general purpose if it is in `@GlobalScope`

But there are already general purpose functions for serialization. The feature which is missing is specifically the workflow with GLSL shaders.

**Comment by AThousandShips** [comment]
Let's take this to the proposal, this is for implementation specific things, and I commented before the proposal was made

**Comment by pkpro** [comment]
> Let's take this to the proposal, this is for implementation specific things, and I commented before the proposal was made

Here is the proposal, let's wait on proposal decision: https://github.com/godotengine/godot-proposals/issues/7478

**Comment by dalexeev** [comment]
I didn't read the discussion, sorry if the questions are repeated.

1. Why is there no inverse function? Note that it will require a `Variant` type parameter to decode.
2. For consistency with other methods, the names should be `a_to_b` and `b_to_a`.
3. If these methods are not going to be used often, I think it makes sense to put them in the `Marshalls` class instead of `@GlobalScope`.

Existing methods:

Return type    |Signature
---------------|------------------------------------------------------------------
PackedByteArray|base64_to_raw(base64_str: String)
String         |base64_to_utf8(base64_str: String)
Variant        |base64_to_variant(base64_str: String, allow_objects: bool = false)
String         |raw_to_base64(array: PackedByteArray)
String         |utf8_to_base64(utf8_str: String)
String         |variant_to_base64(variant: Variant, full_objects: bool = false)

Logical continuation:

Return type    |Signature
---------------|-------------------------------------------------------------------
PackedByteArray|variant_to_raw(variant: Variant, full_objects: bool = false)
Variant        |raw_to_variant(array: PackedByteArray, allow_objects: bool = false)

What could be the methods:

Return type    |Signature
---------------|-------------------------------------------------------------------------------------------
PackedByteArray|variant_to_rawdata(variant: Variant, full_objects: bool = false)
Variant        |rawdata_to_variant(array: PackedByteArray, type: Variant.Type, allow_objects: bool = false)

Note that binary serialization may encode additional data (see [flags](https://github.com/godotengine/godot/blob/4714e95896c8db02616ea6ec7f9aff92dec1cae4/core/io/marshalls.cpp#L58-L60)). For example, we can add information about typed arrays, see #78219.

**Comment by pkpro** [comment]
> I didn't read the discussion, sorry if the questions are repeated.
> 
>     1. Why is there no inverse function? Note that it will require a `Variant` type parameter to decode.
> 

That is the point it is not a serialization method, you can't deserealize from the result. No any metadata available in the raw byte stream. 

>     2. For consistency with other methods, the names should be `a_to_b` and `b_to_a`.
> 

It is impossible IMHO. Prove me wrong please.

>     3. If these methods are not going to be used often, I think it makes sense to put them in the `Marshalls` class instead of `@GlobalScope`.
> 
> 
> Existing methods:
> Return type 	Signature
> PackedByteArray 	base64_to_raw(base64_str: String)
> String 	base64_to_utf8(base64_str: String)
> Variant 	base64_to_variant(base64_str: String, allow_objects: bool = false)
> String 	raw_to_base64(array: PackedByteArray)
> String 	utf8_to_base64(utf8_str: String)
> String 	variant_to_base64(variant: Variant, full_objects: bool = false)
> 
> Logical continuation:
> Return type 	Signature
> PackedByteArray 	variant_to_raw(variant: Variant, full_objects: bool = false)
> Variant 	raw_to_variant(array: PackedByteArray, allow_objects: bool = false)
> 
> What could be the methods:
> Return type 	Signature
> PackedByteArray 	variant_to_rawdata(variant: Variant, full_objects: bool = false)
> Variant 	rawdata_to_variant(array: PackedByteArray, type: Variant.Type, allow_objects: bool = false)
> 
> Note that binary serialization may encode additional data (see [flags](https://github.com/godotengine/godot/blob/4714e95896c8db02616ea6ec7f9aff92dec1cae4/core/io/marshalls.cpp#L58-L60)). For example, we can add information about typed arrays, see #78219.

Serialization is not the scope of this function. The purpose of this function is one way conversion of GDScript Variant (limited subset of it) into uniforms of GLSL compute shaders keeping the bit-wise and semantic conversion at bay.

**Comment by dalexeev** [comment]
> Serialization is not the scope of this function. The purpose of this function is one way conversion of GDScript Variant (limited subset of it) into uniforms of GLSL compute shaders keeping the bit-wise and semantic conversion at bay.

If this is such a specific task, then it should be a `RenderingDevice` method or some `RD*` class, not a global function.

**Comment by pkpro** [comment]
> > Serialization is not the scope of this function. The purpose of this function is one way conversion of GDScript Variant (limited subset of it) into uniforms of GLSL compute shaders keeping the bit-wise and semantic conversion at bay.
> 
> If this is such a specific task, then it should be a `RenderingDevice` method or some `RD*` class, not a global function.

Probably. I can't decide it, I'm new to Ggodot engine. It has been mentioned in proposal. From my point of view, it will be implemented there were it should be and you will tell me where, though there is no decision on this enhancement yet.

**Pull request closed without merging**